qiskit/pulse/builder.py

```diff
@@ -1305,15 +1317,20 @@ def frequency_offset(frequency: float,
         None
     """
     builder = _active_builder()
-    t0 = builder.context_schedule.duration
+    t0 = builder._context_stack[-1].duration
 
     for channel in channels:
         shift_frequency(frequency, channel)
     try:
         yield
     finally:
         if compensate_phase:
-            duration = builder.context_schedule.duration - t0
+            # In some alignment such as sequential and functionl, absolute position of pulse
+            # depends on the number of pulses added after this context.
+            # Thus this logic is not always correct. Use Frame instead of this.
+            warnings.warn('Phase compensation with `compensate_phase=True` is being deprecated. '
+                          'Use Frame instead. See Qiskit-terra/#5977.', DeprecationWarning)
```

Comment:
Maybe we shouldn't add this warning here, but in the signal PR instead

Reply:
Good suggestion. @eggerdj can you add this deprecation warning to #5977 ?

Reply:
Done in 028e9df